p2p/server/server.go

```diff
@@ -73,36 +99,91 @@ type Host interface {
 
 // Server for the Handler.
 type Server struct {
-	logger       log.Log
-	protocol     string
-	handler      Handler
-	timeout      time.Duration
-	requestLimit int
+	logger              log.Log
+	protocol            string
+	handler             Handler
+	timeout             time.Duration
+	requestLimit        int
+	queueSize           int
+	requestsPerInterval int
+	interval            time.Duration
 
-	h Host
+	metrics *tracker // metrics can be nil
 
-	ctx context.Context
+	h Host
 }
 
 // New server for the handler.
 func New(h Host, proto string, handler Handler, opts ...Opt) *Server {
 	srv := &Server{
-		ctx:          context.Background(),
-		logger:       log.NewNop(),
-		protocol:     proto,
-		handler:      handler,
-		h:            h,
-		timeout:      10 * time.Second,
-		requestLimit: 10240,
+		logger:              log.NewNop(),
+		protocol:            proto,
+		handler:             handler,
+		h:                   h,
+		timeout:             10 * time.Second,
+		requestLimit:        10240,
+		queueSize:           1000,
+		requestsPerInterval: 100,
+		interval:            time.Second,
 	}
 	for _, opt := range opts {
 		opt(srv)
 	}
-	h.SetStreamHandler(protocol.ID(proto), srv.streamHandler)
 	return srv
 }
 
-func (s *Server) streamHandler(stream network.Stream) {
+type request struct {
+	stream   network.Stream
+	received time.Time
+}
+
+func (s *Server) Run(ctx context.Context) error {
+	limit := rate.NewLimiter(rate.Every(s.interval), s.requestsPerInterval)
+	queue := make(chan request, s.queueSize)
+	if s.metrics != nil {
+		s.metrics.targetQueue.Set(float64(s.queueSize))
+		s.metrics.targetRps.Set(float64(limit.Limit()))
+	}
+	s.h.SetStreamHandler(protocol.ID(s.protocol), func(stream network.Stream) {
+		select {
+		case queue <- request{stream: stream, received: time.Now()}:
+			if s.metrics != nil {
+				s.metrics.queue.Set(float64(len(queue)))
+				s.metrics.accepted.Inc()
+			}
+		default:
+			if s.metrics != nil {
+				s.metrics.dropped.Inc()
+			}
+			stream.Close()
+		}
+	})
+
+	eg, ctx := errgroup.WithContext(ctx)
+	eg.SetLimit(s.queueSize)
+	for {
+		select {
+		case <-ctx.Done():
+			eg.Wait()
+			return nil
+		case req := <-queue:
+			if err := limit.Wait(ctx); err != nil {
```

Comment:
This `ctx` is not the `ctx` passed to `Run` but the context created from `errgroup.WithContext` which should only be used by go routines managed by `errgroup`.